File: intel_briefing/pipeline/collector.py
# ABOUTME: Parallel sensor fetch coordinator for Intel Briefing.
# ABOUTME: Runs all enabled sensors concurrently, collects results, and writes the cache.
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed, TimeoutError
from datetime import datetime, timezone

from intel_briefing.models import IntelItem, IntelReport, ConfigSettings, SensorResult
from intel_briefing.pipeline.dedup import dedup_items, dedup_across_sections
from intel_briefing.pipeline.cache import write_cache, DEFAULT_CACHE_PATH

logger = logging.getLogger(__name__)

# Per-sensor fetch timeout in seconds
_SENSOR_TIMEOUT = 60


def _load_sensors(config: ConfigSettings) -> list:
    """Dynamically load all enabled sensors based on config toggles."""
    sensors = []
    enabled = config.sensors_enabled

    # Each entry: (sensor_module_path, class_or_factory_name, sensor_key)
    sensor_registry = [
        ("intel_briefing.sensors.hacker_news", "HackerNewsSensor", "hacker_news"),
        ("intel_briefing.sensors.arxiv", "ArxivSensor", "arxiv"),
        ("intel_briefing.sensors.github", "GitHubSensor", "github"),
        ("intel_briefing.sensors.product_hunt", "ProductHuntSensor", "product_hunt"),
        ("intel_briefing.sensors.v2ex", "V2EXSensor", "v2ex"),
        ("intel_briefing.sensors.hn_blogs_sensor", "HNBlogsSensor", "hn_blogs"),
        ("intel_briefing.sensors.grok", "GrokSensor", "grok"),
        ("intel_briefing.sensors.sources_36kr", "Sources36KrSensor", "sources_36kr"),
        ("intel_briefing.sensors.wallstreetcn", "WallStreetCNSensor", "wallstreetcn"),
        ("intel_briefing.sensors.politics", "PoliticsSensor", "politics"),
        ("intel_briefing.sensors.topics", "TopicsSensor", "topics"),
    ]

    for module_path, class_name, key in sensor_registry:
        if not enabled.get(key, True):
            logger.debug("Sensor %s is disabled in config", key)
            continue
        try:
            import importlib
            module = importlib.import_module(module_path)
            sensor_cls = getattr(module, class_name)
            sensors.append(sensor_cls())
        except (ImportError, AttributeError) as exc:
            logger.warning("Could not load sensor %s: %s", key, exc)

    return sensors


def _run_sensor(sensor, config: ConfigSettings, limit: int) -> SensorResult:
    """Run a single sensor's fetch method and return a SensorResult."""
    try:
        items = sensor.fetch(config, limit)
        logger.info("Sensor %s returned %d items", sensor.sensor_name, len(items))
        return SensorResult(sensor_name=sensor.sensor_name, items=items)
    except Exception as exc:
        logger.warning("Sensor %s failed: %s", sensor.sensor_name, exc)
        return SensorResult(sensor_name=sensor.sensor_name, error=str(exc))


# Section routing: maps sensor_name to report section key
_SENSOR_SECTION_MAP: dict[str, str] = {
    "hacker_news": "tech_trends",
    "github": "tech_trends",
    "grok": "tech_trends",
    "arxiv": "research",
    "hn_blogs": "insights",
    "product_hunt": "products",
    "v2ex": "community",
    "sources_36kr": "capital_flow",
    "wallstreetcn": "capital_flow",
    "politics": "politics",
    "topics": "topics",
}


def collect(config: ConfigSettings, cache_path=DEFAULT_CACHE_PATH) -> IntelReport:
    """Run the full collection pipeline and return a structured IntelReport.

    Steps:
    1. Load all enabled sensors.
    2. Fetch from all sensors concurrently with per-sensor timeouts.
    3. Deduplicate items within each section.
    4. Deduplicate across politics / topics sections.
    5. Write the result to the JSON cache.

    Args:
        config: Full application settings.
        cache_path: Where to write the cache file.

    Returns:
        The assembled IntelReport.
    """
    sensors = _load_sensors(config)
    if not sensors:
        logger.warning("No sensors loaded — check sensors_enabled config")

    limit = config.default_limit
    results: list[SensorResult] = []

    with ThreadPoolExecutor(max_workers=min(16, len(sensors) or 1)) as executor:
        future_to_sensor = {
            executor.submit(_run_sensor, sensor, config, limit): sensor
            for sensor in sensors
        }
        for future in as_completed(future_to_sensor, timeout=_SENSOR_TIMEOUT * 2):
            sensor = future_to_sensor[future]
            try:
                result = future.result(timeout=_SENSOR_TIMEOUT)
            except TimeoutError:
                logger.warning("Sensor %s timed out", sensor.sensor_name)
                result = SensorResult(sensor_name=sensor.sensor_name, error="timeout")
            except Exception as exc:
                logger.warning("Sensor %s raised unexpected error: %s", sensor.sensor_name, exc)
                result = SensorResult(sensor_name=sensor.sensor_name, error=str(exc))
            results.append(result)

    # Assemble sections
    sections: dict[str, list[IntelItem]] = {
        "tech_trends": [], "research": [], "capital_flow": [],
        "products": [], "community": [], "politics": [],
        "topics": [], "insights": [],
    }
    sources_ok: list[str] = []
    sources_failed: list[str] = []

    for result in results:
        if result.succeeded:
            sources_ok.append(result.sensor_name)
            section = _SENSOR_SECTION_MAP.get(result.sensor_name, "tech_trends")
            sections[section].extend(result.items)
        else:
            sources_failed.append(result.sensor_name)

    # Deduplicate within each section
    for key in sections:
        sections[key] = dedup_items(sections[key])

    # Deduplicate across politics / topics
    sections = dedup_across_sections(sections)

    now = datetime.now(timezone.utc)
    report = IntelReport(
        date=now.strftime("%Y-%m-%d"),
        fetched_at=now.strftime("%Y-%m-%dT%H:%M:%SZ"),
        stale=False,
        sources_ok=sorted(sources_ok),
        sources_failed=sorted(sources_failed),
        items=sections,
    )

    try:
        write_cache(report, cache_path)
    except Exception as exc:
        logger.error("Failed to write cache: %s", exc)

    return report

File: frontend/src/lib/pipeline/collector.ts
// ABOUTME: Parallel sensor fetch coordinator for Intel Briefing.
// ABOUTME: Runs all enabled sensors concurrently, collects results, and writes the SQLite cache.
import type {
  ConfigSettings,
  IntelItem,
  IntelReport,
  SensorResult,
  SectionKey,
} from '../models'
import { createReport, sensorResultSucceeded, emptyItemsMap, sensorLimit } from '../models'
import { dedupItems, dedupAcrossSections } from './dedup'
import { writeReport } from './cache'
import { SENSOR_REGISTRY } from '../sensors'
import { SensorConfigError } from '../sensors/errors'
import { verifyLink } from '../utils/verifier'
import { fetchContent } from '../utils/jina-reader'

// Section routing: maps sensor_name to report section key
const SENSOR_SECTION_MAP: Record<string, SectionKey> = {
  hacker_news: 'tech_trends',
  github: 'tech_trends',
  arxiv: 'research',
  hn_blogs: 'insights',
  product_hunt: 'products',
  v2ex: 'community',
  sources_36kr: 'capital_flow',
  wallstreetcn: 'capital_flow',
  social_accounts: 'social',
  social_topics: 'social',
  social_trends: 'social',
  chrome_radar: 'products',
}

type ProgressCallback = (
  sensorName: string,
  state: string,
  itemCount: number,
  error: string | null,
  errorKind: 'config' | 'api' | null,
) => void | Promise<void>

/**
 * Run a single sensor's fetch function and return a SensorResult.
 */
async function runSensor(
  sensorName: string,
  fetchFn: (config: ConfigSettings, limit: number) => Promise<IntelItem[]>,
  config: ConfigSettings,
  limit: number,
): Promise<SensorResult> {
  try {
    const items = await fetchFn(config, limit)
    return { sensor_name: sensorName, items, error: null, error_kind: null }
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err)
    const isConfig = err instanceof SensorConfigError
    return { sensor_name: sensorName, items: [], error: message, error_kind: isConfig ? 'config' : 'api' }
  }
}

/**
 * Run the full collection pipeline and return a structured IntelReport.
 *
 * Steps:
 * 1. Identify all enabled sensors.
 * 2. Fetch from all sensors concurrently with Promise.allSettled().
 * 3. Deduplicate items within each section.
 * 4. Deduplicate within the social section (accounts over topics/trends).
 * 5. Write the result to the SQLite cache.
 */
export async function collect(
  config: ConfigSettings,
  onProgress?: ProgressCallback,
): Promise<IntelReport> {
  // Identify enabled sensors
  const enabledSensors = Object.entries(SENSOR_REGISTRY).filter(
    ([name]) => config.sensors_enabled[name] !== false,
  )

  // Mark all as running (sequential awaits to prevent write-ordering races)
  if (onProgress) {
    for (const [name] of enabledSensors) {
      await onProgress(name, 'running', 0, null, null)
    }
  }

  // Run all sensors concurrently, using the per-sensor limit or global default
  const promises = enabledSensors.map(([name, fetchFn]) => {
    const limit = sensorLimit(config, name)
    return runSensor(name, fetchFn, config, limit)
  })
  const settled = await Promise.allSettled(promises)

  const results: SensorResult[] = []
  for (const outcome of settled) {
    if (outcome.status === 'fulfilled') {
      results.push(outcome.value)
    } else {
      // Promise rejected (should be rare since runSensor catches errors)
      results.push({
        sensor_name: 'unknown',
        items: [],
        error: String(outcome.reason),
      })
    }
  }

  // Notify progress for each result (sequential awaits to prevent write-ordering races)
  if (onProgress) {
    for (const result of results) {
      if (sensorResultSucceeded(result)) {
        await onProgress(result.sensor_name, 'ok', result.items.length, null, null)
      } else {
        await onProgress(result.sensor_name, 'failed', 0, result.error, result.error_kind ?? 'api')
      }
    }
  }

  // Assemble sections
  const sections = emptyItemsMap()
  const sourcesOk: string[] = []
  const sourcesFailed: string[] = []

  for (const result of results) {
    if (sensorResultSucceeded(result)) {
      sourcesOk.push(result.sensor_name)
      const section = SENSOR_SECTION_MAP[result.sensor_name] ?? 'tech_trends'
      sections[section].push(...result.items)
    } else {
      sourcesFailed.push(result.sensor_name)
    }
  }

  // Deduplicate within each section
  for (const key of Object.keys(sections) as SectionKey[]) {
    sections[key] = dedupItems(sections[key])
  }

  // Deduplicate within the social section (accounts take priority over topics/trends)
  const dedupedSections = dedupAcrossSections(sections)

  // Post-processing: verify links (Grok items) + enrich content (hn_blogs) — concurrent
  const postProcessTasks: Promise<void>[] = []

  for (const key of Object.keys(dedupedSections) as SectionKey[]) {
    for (const item of dedupedSections[key]) {
      if (item.source === 'x' && item.url) {
        postProcessTasks.push(
          verifyLink(item.url).then(ok => { item.verified = ok }),
        )
      }
      if (item.source === 'hn_blogs' && item.url) {
        postProcessTasks.push(
          fetchContent(item.url).then(text => {
            if (text) item.content = text
          }),
        )
      }
    }
  }

  await Promise.allSettled(postProcessTasks)

  const now = new Date()
  const report = createReport({
    date: now.toISOString().slice(0, 10),
    fetched_at: now.toISOString().replace(/\.\d+Z$/, 'Z'),
    stale: false,
    sources_ok: sourcesOk.sort(),
    sources_failed: sourcesFailed.sort(),
    items: dedupedSections as Record<SectionKey, IntelItem[]>,
  })

  // Write to SQLite cache
  try {
    await writeReport(report)
  } catch (err) {
    console.error('Failed to write cache:', err)
  }

  return report
}
